tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/service/impl/TrainingNumberServiceImpl.java

```diff
@@ -117,4 +136,293 @@ public void delete(Long id) {
+   * Get the training number suffix for the given programme membership.
+   *
+   * @param programmeMembership The programme membership to get the suffix for.
+   * @return The calculated suffix for the programme membership's training number.
+   */
+  private String getSuffix(ProgrammeMembership programmeMembership) {
+    log.info("Calculating suffix.");
+    String trainingPathway = programmeMembership.getTrainingPathway();
+    log.info("Using training pathway '{}' to calculate suffix.", trainingPathway);
+
+    String suffix;
+    switch (trainingPathway) {
+      case "CCT":
+        suffix = "C";
+        break;
+      case "CESR":
+        suffix = "CP";
+        break;
+      default:
+        List<CurriculumDTO> sortedCurricula = filterAndSortCurricula(programmeMembership);
+        String firstSpecialtyCode = sortedCurricula.get(0).getSpecialty().getSpecialtyCode();
+        log.info("Using specialty code '{}' to calculate suffix.", trainingPathway);
+
+        suffix = Objects.equals(firstSpecialtyCode, "ACA") ? "C" : "D";
+    }
+
+    log.info("Calculated suffix: '{}'.", suffix);
+    return suffix;
+  }
+
+  /**
+   * Filter a programme membership's curricula and sort them alphanumerically.
+   *
+   * @param programmeMembership The programme membership to filter and sort the curricula of.
+   * @return The valid curricula for this PM, sorted alphanumerically by subtype and code.
+   */
+  private List<CurriculumDTO> filterAndSortCurricula(ProgrammeMembership programmeMembership) {
+    LocalDate startDate = programmeMembership.getProgrammeStartDate();
+    LocalDate now = LocalDate.now();
+    LocalDate filterDate = startDate.isAfter(now) ? startDate : now;
+
+    Set<String> uniqueSpecialtyCodes = new HashSet<>();
+
+    return programmeMembership.getCurriculumMemberships().stream()
+        .filter(cm -> !cm.getCurriculumStartDate().isAfter(filterDate))
+        .filter(cm -> !cm.getCurriculumEndDate().isBefore(filterDate))
+        .map(cm -> curriculumService.findOne(cm.getCurriculumId()))
+        .filter(c -> c.getSpecialty().getSpecialtyCode() != null && !Strings.isBlank(
+            c.getSpecialty().getSpecialtyCode()))
+        .sorted(Comparator
+            .comparing(CurriculumDTO::getCurriculumSubType)
+            .reversed()
+            .thenComparing(c -> c.getSpecialty().getSpecialtyCode())
+            .reversed()
+        )
+        .filter(c -> uniqueSpecialtyCodes.add(c.getSpecialty().getSpecialtyCode()))
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Check whether the given programme membership is excluded for training number generation.
+   *
+   * @param programmeMembership The programme membership to check.
+   * @return true if training number generated cannot continue, else false.
+   */
+  private boolean isExcluded(ProgrammeMembership programmeMembership) {
+    boolean personExcluded = isExcluded(programmeMembership.getPerson());
+
+    if (personExcluded) {
+      return true;
+    }
```

Comment:
Good spot, seems like a victim of the refactor. It *was* inline in the original TSS implementation but I shuffled things around a few times with the Person record being available under the PM instead of coming from outside.